### ISSUE TITLE
Support custom Hibernate properties in JPADatabase

### ISSUE THREAD

**Issue by jkubrynski** [jOOQ/jOOQ#6709]
Currently AFAIK there is no way of passing custom Hibernate properties to the generator. It's extremely important when you use custom naming strategies:
- `hibernate.physical_naming_strategy`
- `hibernate.implicit_naming_strategy`

and possibly also:
- `hibernate.dialect`

**Comment by lukaseder** [comment]
Thank you very much for your suggestion. In principle, the fact that Hibernate is being used behind the scenes should be transparent to the user of the `JPADatabase`. E.g. we could be replacing that dependency by EclipseLink any time. On the other hand, I do understand the point of using naming strategies.

I'll think about it.

**Comment by jkubrynski** [comment]
If you want to be JPA provider agnostic maybe just allow specifying custom properties that will be passed to the schema generation tool?

**Comment by lukaseder** [comment]
What are "custom properties"? :) I mean, sure we can allow any kind of property to use with the `JPADatabase`, but how can we maintain where these properties belong? I.e. the current implementation of the JPADatabase does this:

```java
Properties info = new Properties();
info.put("user", "sa");
info.put("password", "");
connection = new org.h2.Driver().connect("jdbc:h2:mem:jooq-meta-extensions", info);

MetadataSources metadata = new MetadataSources(
    new StandardServiceRegistryBuilder()
        .applySetting("hibernate.dialect", HIBERNATE_DIALECT)
        .applySetting("javax.persistence.schema-generation-connection", connection)
        .applySetting("javax.persistence.create-database-schemas", true)

        // [#5607] JPADatabase causes warnings - This prevents them
        .applySetting(AvailableSettings.CONNECTION_PROVIDER, connectionProvider())
        .build()
);

ClassPathScanningCandidateComponentProvider scanner =
    new ClassPathScanningCandidateComponentProvider(true);

scanner.addIncludeFilter(new AnnotationTypeFilter(Entity.class));

// [#5845] Use the correct ClassLoader to load the jpa entity classes defined in the user project
ClassLoader cl = Thread.currentThread().getContextClassLoader();

for (String pkg : packages.split(","))
    for (BeanDefinition def : scanner.findCandidateComponents(defaultIfBlank(pkg, "").trim()))
        metadata.addAnnotatedClass(Class.forName(def.getBeanClassName(), true, cl));

// This seems to be the way to do this in idiomatic Hibernate 5.0 API
// See also: http://stackoverflow.com/q/32178041/521799
// SchemaExport export = new SchemaExport((MetadataImplementor) metadata.buildMetadata(), connection);
// export.create(true, true);

// Hibernate 5.2 broke 5.0 API again. Here's how to do this now:
SchemaExport export = new SchemaExport();
export.create(EnumSet.of(TargetType.DATABASE), metadata.buildMetadata());
```

See the slightly snarky comment about Hibernate minor releases breaking API again. Where should we inject what property? To the `StandardServiceRegistryBuilder`? Or the `SchemaExport`? Will this work in Hibernate 6.0, still? Is there a similar API in other JPA providers?

Supporting such properties seems like a maintenance nightmare, when it would be really simple to just copy paste the entire `JPADatabase`: https://github.com/jOOQ/jOOQ/blob/master/jOOQ-meta-extensions/src/main/java/org/jooq/util/jpa/JPADatabase.java and roll your own implementation.

Again, I don't deny that your suggestion is perfectly valid and we should find a way to support such features, but it is hard to get right, I'm afraid, thus the above workaround of patching this manually for the time being.

**Comment by jkubrynski** [comment]
I think the most important properties are related to the `MetadataSources`. That's the place where you set dialect, naming strategies, etc. Connection properties are more related to the way Jooq generates the schema. I'm not saying it's a best solution but please take a look here: https://github.com/Devskiller/hbm2ddl-maven-plugin/blob/master/src/main/java/com/devskiller/hbm2ddl/SchemaGenerator.java#L36

Also I'm pretty sure that Hibernate won't remove a way to set custom properties

**Comment by lukaseder** [comment]
Yeah, sure, that's one way to do it

> Also I'm pretty sure that Hibernate won't remove a way to set custom properties

I'd say they certainly won't remove the property. But quite possibly, they might rename it, or set it somewhere else. The background of my fear is here: http://stackoverflow.com/q/32178041/521799

But again, I agree with the suggestion as a thing that jOOQ should support. I'm just not sure about how to best implement it yet. In the worst case, we can write some hard-wired logic into the `JPADatabase` that routes each "known property" to the proper place. That's more or less what most of Spring's glue code does. I could live with that.

**Comment by jkubrynski** [comment]
The main question is if you want to do it separately for naming strategies, dialect, etc. Then you can even ignore properties and configure it directly on a `MetadataBuilder`. Rewriting properties from configuration is just a simpler idea :)

**Comment by lukaseder** [comment]
I still don't have a well engineered solution for this. The `JPADatabase` is only 250 lines of code. As it stands now, and until there is more traction to this feature request, I think the default solution will be to fork the `JPADatabase` and apply the settings manually.

**Comment by XhstormR** [comment]
I also need this feature.
When I use to generate code from the database, the code looks like this:
```java
public final TableField<AccountRecord, LocalDateTime> CREATED_AT = createField("created_at", org.jooq.impl.SQLDataType.LOCALDATETIME.nullable(false).defaultValue(org.jooq.impl.DSL.field("CURRENT_TIMESTAMP", org.jooq.impl.SQLDataType.LOCALDATETIME)), this, "");
```
But when I use the code generated from the JPA entity, the generated code is like this:
```java
public final TableField<AccountRecord, LocalDateTime> CREATEDAT = createField("CREATEDAT", org.jooq.impl.SQLDataType.LOCALDATETIME.nullable(false).defaultValue(org.jooq.impl.DSL.field("CURRENT_TIMESTAMP", org.jooq.impl.SQLDataType.LOCALDATETIME)), this, "");
```
The field name in the generated code does not match the previous one. If you do not modify the source code used, the compilation will fail. And the readability of the field name is not good.

**Comment by lukaseder** [comment]
> But when I use the code generated from the JPA entity, the generated code is like this:

What does the entity look like?

**Comment by XhstormR** [comment]
This is the code for BaseEntity: 
```kotlin
@MappedSuperclass
abstract class BaseEntity(
        @Version
        @Column(nullable = false, name = "OPTLOCK")
        var version: Long = 0,
        @Column(nullable = false, insertable = false, updatable = false)
        var createdAt: LocalDateTime = LocalDateTime.now(),
        @Column(nullable = false, insertable = false, updatable = true)
        var updatedAt: LocalDateTime = LocalDateTime.now()
) :  Serializable {
    companion object {
        private const val serialVersionUID = 1L
    }
}
```
I wrote a test Kotlin code based on the code of JPADatabase, which can generate attribute names underlined as shown below:
```kotlin
    fun test() {
        val metadata = MetadataSources(
                StandardServiceRegistryBuilder()
                        .applySetting("hibernate.dialect", "org.hibernate.dialect.PostgreSQL95Dialect")
                        .applySetting("hibernate.default_schema", "erp")
                        .applySetting("javax.persistence.create-database-schemas", true)
                        .applySetting("hibernate.physical_naming_strategy", "org.springframework.boot.orm.jpa.hibernate.SpringPhysicalNamingStrategy")
                        .build()
        )

        val scanner = ClassPathScanningCandidateComponentProvider(false).apply {
            addIncludeFilter(AnnotationTypeFilter(Entity::class.java))
        }

        for (bean in scanner.findCandidateComponents("com.xhstormr.erp.domain.model.entity")) {
            metadata.addAnnotatedClass(Class.forName(bean.beanClassName))
        }

        val export = SchemaExport().apply {
            setDelimiter(";")
            setOutputFile("create.sql")
        }
        export.createOnly(EnumSet.of(TargetType.SCRIPT), metadata.buildMetadata())
    }
```

The key code is this:
```
.applySetting("hibernate.physical_naming_strategy", "org.springframework.boot.orm.jpa.hibernate.SpringPhysicalNamingStrategy")
```
By default hibernate generates the same field name as the property name, Since I am using Spring boot, it [automatically configures](https://github.com/spring-projects/spring-boot/commit/0db11940078a7a4d946af3f8f28d0d6dfe4d1bb1) the hibernate naming convention to generate more readable field names. 

It would be nice if can add custom attributes [here](https://github.com/jOOQ/jOOQ/blob/master/jOOQ-meta-extensions/src/main/java/org/jooq/meta/extensions/jpa/JPADatabase.java#L131).Now my workaround is to write the generated SQL script first, then use DDLDatabase to generate the database.

**Comment by lukaseder** [comment]
Thanks a lot for your help.

> Now my workaround is to write the generated SQL script first, then use DDLDatabase to generate the database.

Of course, my suggestion of copy pasting the `JPADatabase` is another option. It really isn't that much code.

**Comment by lukaseder** [comment]
OK, perhaps we can be somewhat pragmatic here. All `JPADatabase` properties that start with `hibernate.` or `javax.persistence.` will be automatically passed to these `applySetting()` calls and possibly override the defaults specified by jOOQ.

We'll add a warning to the documentation that the behaviour of this may change in the future, depending on:

- Hibernate changing this behaviour
- We possibly no longer using Hibernate behind the scenes

**Comment by lukaseder** [comment]
This is implemented for jOOQ 3.12.
